services/app-api/src/handlers/bulk_download.ts

```diff
@@ -5,257 +5,288 @@ import {
+                    message: 'No body found in request',
+                }),
+                headers: {
+                    'Access-Control-Allow-Origin': '*',
+                    'Access-Control-Allow-Credentials': true,
+                },
+            }
+        }
+
+        const bulkDlRequest: S3BulkDownloadRequest = JSON.parse(event.body)
+        console.info('Bulk download request:', bulkDlRequest)
+
+        if (
+            !bulkDlRequest.bucket ||
+            !bulkDlRequest.keys ||
+            !bulkDlRequest.zipFileName
+        ) {
+            return {
+                statusCode: 400,
+                body: JSON.stringify({
+                    code: 'BAD_REQUEST',
+                    message: 'Missing required fields',
+                }),
+                headers: {
+                    'Access-Control-Allow-Origin': '*',
+                    'Access-Control-Allow-Credentials': true,
+                },
+            }
+        }
 
-                    // this is the node stream we will return as a Reader.
-                    const passThrough = new PassThrough()
+        // Start memory logging
+        memoryLoggingInterval = startMemoryLogging()
 
-                    // Get a web stream we can write to. Unfortunately the types don't match enough to just transform our StreamReader
-                    const webWriter = Writable.toWeb(passThrough)
+        // Prepare temp directory
+        ensureTempDir()
 
-                    // do we need to await this?
-                    void readStreamBody.pipeTo(webWriter)
+        // Download files in batches
+        let totalBytes = 0
+        const downloadedFiles: DownloadedFile[] = []
 
-                    return {
-                        stream: passThrough,
-                        filename,
-                    }
-                } else {
-                    console.error(
-                        'Programming Error: Unknown return type for s3 Body',
-                        s3Item.Body
-                    )
+        for (let i = 0; i < bulkDlRequest.keys.length; i += BATCH_SIZE) {
+            const batch = bulkDlRequest.keys.slice(i, i + BATCH_SIZE)
+            console.info(
+                `Processing batch ${i / BATCH_SIZE + 1}/${Math.ceil(bulkDlRequest.keys.length / BATCH_SIZE)}`
+            )
+
+            const batchResults = await Promise.all(
+                batch.map(async (key) => {
+                    const timeout = BASE_TIMEOUT + TIMEOUT_PER_MB * 1000 // Simplified timeout calculation
+                    return downloadFile(bulkDlRequest.bucket, key, timeout)
+                })
+            )
+
+            for (const result of batchResults) {
+                totalBytes += result.size
+                if (totalBytes > MAX_TOTAL_SIZE) {
                     throw new Error(
-                        'Programming Error: Unknown return type for s3 Body'
+                        `Total size exceeds maximum allowed size of ${MAX_TOTAL_SIZE} bytes`
```

Comment:
Is this the error we are expecting to see if this issues happens again?  I'd love to leave this work with a clear error thrown instead of things silently exiting if we are going to hit this again

Reply:
I still don't know what the error path of the original failure was actually. Even after I added a bunch of verbosity to what was going on, the lambda would still just shut itself off after zipping a fraction of the files.

This size check is just what the default temp storage size is on a lambda environment. We clean up after ourselves, but basically we don't want to be creating zip archives of > 500MB. We can tweak this value though, I'm up for suggestions.

Oh, I guess this is the closed PR actually 😂